### PR TITLE
[WIP] Export Faster RCNN to ONNX

### PR DESCRIPTION
**Pull request by BowenBao**
Based on [existing work](https://github.com/facebookresearch/maskrcnn-benchmark/pull/138) of enabling Tracing/Scripting for MaskRCNN/FasterRCNN models, this branch extends the support for exporting FasterRCNN to ONNX.

With this patch, we are able to export [e2e_faster_rcnn_R-50-FPN_1x](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/configs/caffe2/e2e_faster_rcnn_R_50_FPN_1x_caffe2.yaml) to ONNX, and run in ONNXRuntime backend.

## Setup Environment

Most parts are the same with [INSTALL.md](../INSTALL.md). We want to use pytorch-1.2 nightly with latest torchvision, so we build torchvision from source.

```bash
conda create --name maskrcnn_onnx
conda activate maskrcnn_onnx
conda install ipython

# maskrcnn_benchmark and coco api dependencies
pip install ninja yacs cython matplotlib tqdm opencv-python requests

# onnx
pip install onnx

# onnxruntime as onnx backend
pip install onnxruntime

# follow PyTorch installation in https://pytorch.org/get-started/locally/
# we give the instructions for CUDA 10.0
conda install -c pytorch pytorch-nightly cudatoolkit=10.0
export INSTALL_DIR=$PWD

# install torchvision
cd $INSTALL_DIR
git clone https://github.com/pytorch/vision.git
cd vision
python setup.py develop

# install pycocotools
cd $INSTALL_DIR
git clone https://github.com/cocodataset/cocoapi.git
cd cocoapi/PythonAPI
python setup.py build_ext install

# install apex
cd $INSTALL_DIR
git clone https://github.com/NVIDIA/apex.git
cd apex
python setup.py install --cuda_ext --cpp_ext

# install PyTorch Detection - ONNX branch
cd $INSTALL_DIR
git clone https://github.com/bowenbao/maskrcnn-benchmark.git maskrcnn-onnx
cd maskrcnn-onnx
git checkout onnx_stage

# the following will install the lib with
# symbolic links, so that you can modify
# the files if you want and won't need to
# re-build it
python setup.py build develop
unset INSTALL_DIR
```

## Export & Inference in ONNX
Under folder [demo](https://github.com/BowenBao/maskrcnn-benchmark/tree/onnx_stage/demo), there are two python demos.

* demo/export_to_onnx.py:
  - Usage: python export_to_onnx.py
  - Exports [e2e_faster_rcnn_R-50-FPN_1x](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/configs/caffe2/e2e_faster_rcnn_R_50_FPN_1x_caffe2.yaml) to faster_rcnn_R_50_FPN_1x.onnx.
* demo/eval_onnx.py:
  - Usage: python eval_onnx.py
  - Loads the exported ONNX model in previous section, and inference with ONNXRuntime.

## TODOs
* This branch depends on many updates to PyTorch ONNX exporter, many are still in PR under review. Once they are in PyTorch master we can remove the ```pytorch_export_patch.py``` patch.
* Add support for other configs and MaskRCNN models.
* To reduce dependencies I commented out the part in custom ops that depends on opencv. This can be added back after clean up setup.py.

### CODE REVIEW

File: maskrcnn_benchmark/csrc/custom_ops/custom_ops.cpp [hunk 1]
```diff
@@ -15,35 +16,35 @@ Tensor upsample_bilinear(const Tensor& inp, int64_t w, int64_t h) {
...
-  cv::Mat cv_res(res.size(0), res.size(1), CV_8UC3, (void*) res.data<uint8_t>());
-  for (int64_t i = 0; i < labels.size(0); i++) {
-    std::stringstream text;
-    text.precision(2);
-    text << class_names[labels[i]] << ": " << scores[i];
-    putText(cv_res, text.str(), cv::Point(bboxes[i][0], bboxes[i][1]), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(color[0], color[1], color[2]), 1);
-  }
-  return res;
-}
+// Tensor add_annotations(const Tensor& image, const Tensor& labels_, const Tensor& scores_, const Tensor& bboxes_, std::string class_names_string, const Tensor& color_) {
```
Comment (t-vi) [inline maskrcnn_benchmark/csrc/custom_ops/custom_ops.cpp:19]: He! I want these for my export... If you introduce something that makes it optional based on some flag, that'd be OK, but if opencv is available, I like to have it by default.

File: demo/coco_classes.txt [hunk 2]
```diff
@@ -0,0 +1,81 @@
+__background
```
Comment (t-vi) [inline demo/coco_classes.txt:1]: This should be independent of the tracing / export.

File: maskrcnn_benchmark/modeling/box_coder.py [hunk 3]
```diff
@@ -70,8 +70,8 @@ def decode(self, rel_codes, boxes):
         TO_REMOVE = 1  # TODO remove
         widths = boxes[:, 2] - boxes[:, 0] + TO_REMOVE
         heights = boxes[:, 3] - boxes[:, 1] + TO_REMOVE
-        ctr_x = boxes[:, 0] + 0.5 * widths
-        ctr_y = boxes[:, 1] + 0.5 * heights
+        ctr_x = boxes[:, 0] + torch.tensor(0.5, dtype=torch.float32) * widths
```
Comment (t-vi) [inline maskrcnn_benchmark/modeling/box_coder.py:73]: This isn't good for general consumption. Are those needed? I cannot really imagine you cannot multiply by a scalar in ONNX.

File: maskrcnn_benchmark/modeling/poolers.py [hunk 4]
```diff
@@ -55,7 +70,7 @@ def __call__(self, boxlists):
         s = torch.sqrt(cat([boxlist.area() for boxlist in boxlists]))
 
         # Eqn.(1) in FPN paper
-        target_lvls = torch.floor(self.lvl0 + torch.log2(s / self.s0 + self.eps))
+        target_lvls = torch.floor(self.lvl0 + torch.log2(torch.tensor(self.eps, dtype=torch.float32) + s / self.s0))
```
Comment (t-vi) [inline maskrcnn_benchmark/modeling/poolers.py:73]: same as above.

File: maskrcnn_benchmark/modeling/rpn/anchor_generator.py [hunk 5]
```diff
@@ -77,13 +77,13 @@ def grid_anchors(self, grid_sizes):
...
             device = base_anchors.device
-            shifts_x = torch.arange(
-                0, grid_width * stride, step=stride, dtype=torch.float32, device=device
-            )
-            shifts_y = torch.arange(
-                0, grid_height * stride, step=stride, dtype=torch.float32, device=device
-            )
-            shift_y, shift_x = torch.meshgrid(shifts_y, shifts_x)
+
+            # WORKAROUND: torch.arange export to ONNX not supported yet.
```
Comment (t-vi) [inline maskrcnn_benchmark/modeling/rpn/anchor_generator.py:81]: Is there a PR or so to fix this? I'd aim for the cleanest solution working with master / master + open PRs.

File: maskrcnn_benchmark/modeling/rpn/inference.py [hunk 6]
```diff
@@ -50,6 +50,11 @@ def __init__(
         self.fpn_post_nms_top_n = fpn_post_nms_top_n
         self.fpn_post_nms_per_batch = fpn_post_nms_per_batch
 
+        self.onnx_export = False
```
Comment (t-vi) [inline maskrcnn_benchmark/modeling/rpn/inference.py:53]: could one make onnx_export a global flag easily? (or maybe @fmassa has an opinion)

File: setup.py [hunk 7]
```diff
@@ -33,7 +34,7 @@ def get_extensions():
                           os.path.join(extensions_dir, "cpu", "ROIAlign_cpu.cpp")]
     custom_ops_sources_cuda = [os.path.join(extensions_dir, "cuda", "nms.cu"),
                                os.path.join(extensions_dir, "cuda", "ROIAlign_cuda.cu")]
-    custom_ops_libraries = ["opencv_core", "opencv_imgproc", "opencv_imgcodecs"]
+    custom_ops_libraries = []#["opencv_core", "opencv_imgproc", "opencv_imgcodecs"]
```
Comment (t-vi) [inline setup.py:37]: can you make this conditional on some env variable being set, please? I'd prefer opencv by default.

File: setup.py [hunk 8]
```diff
@@ -77,10 +78,13 @@ def get_extensions():
 class rename_custom_ops_lib(distutils.command.build.build):
     def run(self):
         inst_dir = os.path.join(self.build_lib, 'maskrcnn_benchmark', 'lib')
+        lib_dir = os.path.join('maskrcnn_benchmark', 'lib')
         lib_suffix = os.path.basename(torch._C.__file__).split('.', 1)[1]  # there must be a better way
         ext = lib_suffix.rsplit('.', 1)[1]
         os.rename(os.path.join(inst_dir, 'custom_ops.' + lib_suffix),
                   os.path.join(inst_dir, 'libmaskrcnn_benchmark_customops.' + ext))
+        shutil.copy(os.path.join(inst_dir, 'libmaskrcnn_benchmark_customops.' + ext),
```
Comment (t-vi) [inline setup.py:86]: While PyTorch does it, dumping stuff into the main tree (without `setup.py develop`) is a terrible habit. Could we please get by without?

**Review by t-vi — COMMENTED** [review]
Thank you for the patch! That's awesome!
I have two concerns and a question:
- Ideally one would make the "upstream PR" smaller rather than larger, e.g. by splitting out things.
- I'm quite doubtful if the tracing / scripting patch will ever be merged.
- Are you subscribed to the issue to field questions?
(In addition to the other review comments, the jpg should be independent from export.)
@fmassa might have more insights.

**Comment by clw5180** [comment]
> Based on [existing work](https://github.com/facebookresearch/maskrcnn-benchmark/pull/138) of enabling Tracing/Scripting for MaskRCNN/FasterRCNN models, this branch extends the support for exporting FasterRCNN to ONNX.
> 
> With this patch, we are able to export [e2e_faster_rcnn_R-50-FPN_1x](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/configs/caffe2/e2e_faster_rcnn_R_50_FPN_1x_caffe2.yaml) to ONNX, and run in ONNXRuntime backend.
> ## Setup Environment
> 
> Most parts are the same with [INSTALL.md](../INSTALL.md). We want to use pytorch-1.2 nightly with latest torchvision, so we build torchvision from source.
> 
> ```shell
> conda create --name maskrcnn_onnx
> conda activate maskrcnn_onnx
> conda install ipython
> 
> # maskrcnn_benchmark and coco api dependencies
> pip install ninja yacs cython matplotlib tqdm opencv-python requests
> 
> # onnx
> pip install onnx
> 
> # onnxruntime as onnx backend
> pip install onnxruntime
> 
> # follow PyTorch installation in https://pytorch.org/get-started/locally/
> # we give the instructions for CUDA 10.0
> conda install -c pytorch pytorch-nightly cudatoolkit=10.0
> export INSTALL_DIR=$PWD
> 
> # install torchvision
> cd $INSTALL_DIR
> git clone https://github.com/pytorch/vision.git
> cd vision
> python setup.py develop
> 
> # install pycocotools
> cd $INSTALL_DIR
> git clone https://github.com/cocodataset/cocoapi.git
> cd cocoapi/PythonAPI
> python setup.py build_ext install
> 
> # install apex
> cd $INSTALL_DIR
> git clone https://github.com/NVIDIA/apex.git
> cd apex
> python setup.py install --cuda_ext --cpp_ext
> 
> # install PyTorch Detection - ONNX branch
> cd $INSTALL_DIR
> git clone https://github.com/bowenbao/maskrcnn-benchmark.git maskrcnn-onnx
> cd maskrcnn-onnx
> git checkout onnx_stage
> 
> # the following will install the lib with
> # symbolic links, so that you can modify
> # the files if you want and won't need to
> # re-build it
> python setup.py build develop
> unset INSTALL_DIR
> ```
> 
> ## Export & Inference in ONNX
> 
> Under folder [demo](https://github.com/BowenBao/maskrcnn-benchmark/tree/onnx_stage/demo), there are two python demos.
> 
>     * demo/export_to_onnx.py:
>       
>       * Usage: python export_to_onnx.py
>       * Exports [e2e_faster_rcnn_R-50-FPN_1x](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/configs/caffe2/e2e_faster_rcnn_R_50_FPN_1x_caffe2.yaml) to faster_rcnn_R_50_FPN_1x.onnx.
> 
>     * demo/eval_onnx.py:
>       
>       * Usage: python eval_onnx.py
>       * Loads the exported ONNX model in previous section, and inference with ONNXRuntime.
> 
> 
> ## TODOs
> 
>     * This branch depends on many updates to PyTorch ONNX exporter, many are still in PR under review. Once they are in PyTorch master we can remove the `pytorch_export_patch.py` patch.
> 
>     * Add support for other configs and MaskRCNN models.
> 
>     * To reduce dependencies I commented out the part in custom ops that depends on opencv. This can be added back after clean up setup.py.

Have you met the error 'RuntimeError: ONNX symbolic expected a constant value in the trace ?' Thanks a lot !

**Comment by Chanbluky** [comment]
I followed the steps that you provided and try to generate faster rcnn ONNX model, but it failed. i got following info:
/opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/modeling/roi_heads/box_head/inference.py:69: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  (torch.tensor([self.detections_per_img], dtype=torch.long),
/opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/modeling/roi_heads/box_head/inference.py:73: TracerWarning: Converting a tensor to a Python integer might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  scores, number_to_keep, dim=0, sorted=True)
Segmentation fault (core dumped)

and i also saved the log as follows:
  %6554 : Long(83) = onnx::Cast[to=7](%6553), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:220:0
  %6555 : Float(83, 4) = onnx::Gather[axis=0](%6542, %6554), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:220:0
  %6556 : Long(83) = onnx::Cast[to=7](%6554), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
  %6557 : Float(83) = onnx::Gather[axis=0](%6543, %6556), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
  %6558 : Long(83) = onnx::Cast[to=7](%6556), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
  %6559 : Long(83) = onnx::Gather[axis=0](%6544, %6558), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
  return (%6555, %6559, %6557)
~  

could you help to debug this issue? the only difference is that i am using cpu only mode.

**Comment by adizhol** [comment]
has anyone tried exporting DeformConvFunction layer (dcn) to onnx?

**Comment by t-vi** [comment]
Heya! This is an usupported fork. @BowenBao 1) Thank you for your patch! 2) if you are interested in keeping this maintained, please do create your own branch. It'll be useful to people.

**Comment by hznuhisejeffrey** [comment]
@BowenBao  hello，did you try to accelerate your onnx model with TensorRT？

**Comment by zf921022** [comment]
> I followed the steps that you provided and try to generate faster rcnn ONNX model, but it failed. i got following info:
> /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/modeling/roi_heads/box_head/inference.py:69: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
> (torch.tensor([self.detections_per_img], dtype=torch.long),
> /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/modeling/roi_heads/box_head/inference.py:73: TracerWarning: Converting a tensor to a Python integer might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
> scores, number_to_keep, dim=0, sorted=True)
> Segmentation fault (core dumped)
> 
> and i also saved the log as follows:
> %6554 : Long(83) = onnx::Cast[to=7](%6553), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:220:0
> %6555 : Float(83, 4) = onnx::Gather[axis=0](%6542, %6554), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:220:0
> %6556 : Long(83) = onnx::Cast[to=7](%6554), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
> %6557 : Float(83) = onnx::Gather[axis=0](%6543, %6556), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
> %6558 : Long(83) = onnx::Cast[to=7](%6556), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
> %6559 : Long(83) = onnx::Gather[axis=0](%6544, %6558), scope: FRCNNModel/GeneralizedRCNN/CombinedROIHeads[roi_heads]/ROIBoxHead[box]/PostProcessor[post_processor] # /opt/samba/nxf07490/anaconda/maskrcnn-onnx/maskrcnn_benchmark/structures/bounding_box.py:222:0
> return (%6555, %6559, %6557)
> ~
> 
> could you help to debug this issue? the only difference is that i am using cpu only mode.

I met the same issue，did you find the reason?

**Pull request closed without merging**